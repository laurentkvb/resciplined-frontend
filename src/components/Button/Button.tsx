import React from "react";

import { ButtonStyled, Icon } from "./button.styles";

interface Props {
  children: React.ReactNode;
  leadingIcon?: string;
  onClick: Function;
  variant?: string;
}

const Button = ({ children, leadingIcon, onClick, variant = "primary", ...props }: Props) => (
  <ButtonStyled
    variant={variant}
    type="submit"
    onClick={() => onClick()}
    fontWeight="bold"
    fontSize={["19px", "19px"]}
    border={["1px solid"]}
    borderColor={["pink"]}
    backgroundColor={["black"]}
    borderRadius="44px"
    color="white"
    height="55px"
    textAlign="center"
    width={[170]}
    {...props}
  >
    {leadingIcon && (
      <Icon fontSize={2} marginRight={3} className={leadingIcon} />
    )}
    {children}
  </ButtonStyled>
);

export default Button;
