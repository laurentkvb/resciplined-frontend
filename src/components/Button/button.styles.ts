import styled from "styled-components";

import {
  background,
  BackgroundProps,
  border,
  BorderProps,
  color,
  ColorProps,
  layout,
  LayoutProps,
  shadow,
  ShadowProps,
  space,
  SpaceProps,
  typography,
  TypographyProps,
  variant
} from "styled-system";

type SystemProps = BackgroundProps &
BorderProps &
ColorProps &
LayoutProps &
ShadowProps &
SpaceProps &
TypographyProps;

interface ButtonProps extends SystemProps {
  variant?: string
}

type IconProps = SpaceProps & TypographyProps;

export const ButtonStyled = styled.button<ButtonProps>`
  box-sizing: border-box;
  min-width: 0;
  transition: transform .3s ease-in-out;

  ${variant({
    variants: {
      primary: {
        backgroundColor: "primary",
        color: "white",
        fontSize: 1
      },
      tertiary: {
        backgroundColor: "tertiary",
        color: "white",
        fontSize: 1
      },
      text: {
        backgroundColor: "transparent",
        fontSize: 2
      }
    }
  })}

  &:hover {
    transform: scale(1.1);
  }

  &&& {
    ${background}
    ${border}
    ${color}
    ${layout}
    ${shadow}
    ${space}
    ${typography}
  }
`;

export const Icon = styled.i<IconProps>`
  &&& {
    ${space}
    ${typography}
  }
`;
