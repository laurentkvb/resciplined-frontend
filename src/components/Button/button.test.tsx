import React from "react";
import { fireEvent, render } from "../../tests/test-utils";
import Button from "./Button";

const onClick = jest.fn();

test("Button primary variant snapshot test", () => {
  const { container } = render(
    <Button onClick={() => onClick()}>Primary button</Button>,
  );

  // Asserts Button component matches snapshot.
  expect(container.firstChild).toMatchSnapshot();
});

test("Button tertiary variant snapshot test", () => {
  const { container } = render(
    <Button variant="tertiary" onClick={() => onClick()}>
      Tertiary button
    </Button>,
  );

  // Asserts Button component matches snapshot.
  expect(container.firstChild).toMatchSnapshot();
});

test("Button text variant snapshot test", () => {
  const { container } = render(
    <Button variant="text" onClick={() => onClick()}>
      Text button
    </Button>,
  );

  // Asserts Button component matches snapshot.
  expect(container.firstChild).toMatchSnapshot();
});

test("Button with leading icon snapshot test", () => {
  const { container } = render(
    <Button onClick={() => onClick()} leadingIcon="fas fa-alien">
      Text button
    </Button>,
  );

  // Asserts Button component matches snapshot.
  expect(container.firstChild).toMatchSnapshot();
});

test("Button with children rendered correctly.", () => {
  const { getByTestId } = render(
    <Button data-testid="Button" onClick={() => onClick()}>
      Click Me!
    </Button>,
  );
  const renderedButton = getByTestId("Button");

  // Asserts Button component is of type button
  expect(renderedButton.tagName).toBe("BUTTON");

  // Asserts Button component renders children correctly.
  expect(renderedButton.textContent).toBe("Click Me!");
});

test("Button with onClick event fired correctly", () => {
  const { getByTestId } = render(
    <Button data-testid="Button" onClick={() => onClick()}>
      Click Test
    </Button>,
  );

  const renderedButton = getByTestId("Button");

  // Asserts Button click event is fired correctly.
  fireEvent.click(renderedButton);
  expect(onClick).toHaveBeenCalledTimes(1);
});

test("Button with leading icon renders icon", () => {
  const { getByTestId } = render(
    <Button
      data-testid="Button"
      onClick={() => onClick()}
      leadingIcon="fas fa-alien"
    >
      Click me!
    </Button>,
  );

  const renderedButton = getByTestId("Button");
  const icon = renderedButton.querySelector("i");

  // Asserts icon is of type I
  expect(icon?.tagName).toBe("I");
});
