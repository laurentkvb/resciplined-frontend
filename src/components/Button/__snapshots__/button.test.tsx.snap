// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button primary variant snapshot test 1`] = `
.c0 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: -webkit-transform .3s ease-in-out;
  -webkit-transition: transform .3s ease-in-out;
  transition: transform .3s ease-in-out;
  background-color: #002880;
  color: #FFFFFF;
  font-size: 0.875rem;
}

.c0:hover {
  -webkit-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
}

.c0.c0.c0 {
  border-radius: 3px;
  color: #FFFFFF;
  height: 40px;
  padding-left: 16px;
  padding-right: 16px;
  font-weight: 500;
  text-align: center;
}

<button
  class="c0"
  color="white"
  font-weight="display"
  height="40px"
  type="submit"
>
  Primary button
</button>
`;

exports[`Button tertiary variant snapshot test 1`] = `
.c0 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: -webkit-transform .3s ease-in-out;
  -webkit-transition: transform .3s ease-in-out;
  transition: transform .3s ease-in-out;
  background-color: #f29323;
  color: #FFFFFF;
  font-size: 0.875rem;
}

.c0:hover {
  -webkit-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
}

.c0.c0.c0 {
  border-radius: 3px;
  color: #FFFFFF;
  height: 40px;
  padding-left: 16px;
  padding-right: 16px;
  font-weight: 500;
  text-align: center;
}

<button
  class="c0"
  color="white"
  font-weight="display"
  height="40px"
  type="submit"
>
  Tertiary button
</button>
`;

exports[`Button text variant snapshot test 1`] = `
.c0 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: -webkit-transform .3s ease-in-out;
  -webkit-transition: transform .3s ease-in-out;
  transition: transform .3s ease-in-out;
  background-color: transparent;
  font-size: 1.125rem;
}

.c0:hover {
  -webkit-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
}

.c0.c0.c0 {
  border-radius: 3px;
  color: #FFFFFF;
  height: 40px;
  padding-left: 16px;
  padding-right: 16px;
  font-weight: 500;
  text-align: center;
}

<button
  class="c0"
  color="white"
  font-weight="display"
  height="40px"
  type="submit"
>
  Text button
</button>
`;

exports[`Button with leading icon snapshot test 1`] = `
.c0 {
  box-sizing: border-box;
  min-width: 0;
  -webkit-transition: -webkit-transform .3s ease-in-out;
  -webkit-transition: transform .3s ease-in-out;
  transition: transform .3s ease-in-out;
  background-color: #002880;
  color: #FFFFFF;
  font-size: 0.875rem;
}

.c0:hover {
  -webkit-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
}

.c0.c0.c0 {
  border-radius: 3px;
  color: #FFFFFF;
  height: 40px;
  padding-left: 16px;
  padding-right: 16px;
  font-weight: 500;
  text-align: center;
}

.c1.c1.c1 {
  margin-right: 12px;
  font-size: 1.125rem;
}

<button
  class="c0"
  color="white"
  font-weight="display"
  height="40px"
  type="submit"
>
  <i
    class="c1 fas fa-alien"
    font-size="2"
  />
  Text button
</button>
`;
